controllers/extendeddaemonsetreplicaset/strategy/canary.go

```diff
@@ -1,98 +1,275 @@
-	autoPauseEnabled := *daemonset.Spec.Strategy.Canary.AutoPause.Enabled
-	maxRestarts := int(*daemonset.Spec.Strategy.Canary.AutoPause.MaxRestarts)
+		needRequeue            bool
+		podsToCheckForRestarts []*v1.Pod
 
-	// Canary mode
+		podsToCreate []*NodeItem
+		podsToDelete []*NodeItem
+	)
+
+	// First scan canary node list for pods to create or delete
 	for _, nodeName := range params.CanaryNodes {
 		node := params.NodeByName[nodeName]
 		desiredPods++
 		if pod, ok := params.PodByNodeName[node]; ok {
 			if pod == nil {
-				result.PodsToCreate = append(result.PodsToCreate, node)
+				podsToCreate = append(podsToCreate, node)
+				continue
+			}
+
+			if pod.DeletionTimestamp != nil {
 				needRequeue = true
-			} else {
-				if err = addPodLabel(client, pod, v1alpha1.ExtendedDaemonSetReplicaSetCanaryLabelKey, v1alpha1.ExtendedDaemonSetReplicaSetCanaryLabelValue); err != nil {
-					params.Logger.Error(err, fmt.Sprintf("Couldn't add the canary label for pod '%s/%s', will retry later", pod.GetNamespace(), pod.GetName()))
-					needRequeue = true
-				}
-				if pod.DeletionTimestamp != nil {
-					needRequeue = true
-					continue
-				}
-				if !compareCurrentPodWithNewPod(params, pod, node) && pod.DeletionTimestamp == nil {
-					result.PodsToDelete = append(result.PodsToDelete, node)
-				} else {
-					currentPods++
-					if podUtils.IsPodAvailable(pod, 0, metaNow) {
-						availablePods++
-					}
-					if podUtils.IsPodReady(pod) {
-						readyPods++
-					}
-					if autoPauseEnabled && !isPaused {
-						// Check if deploy should be paused due to restarts. Note that pausing the canary will have no effect if it has been validated or failed
-						if isRestarting, reason := podUtils.IsPodRestarting(pod, maxRestarts); isRestarting {
-							err = pauseCanaryDeployment(client, daemonset, reason)
-							if err != nil {
-								params.Logger.Error(err, "Failed to pause canary deployment")
-							} else {
-								params.Logger.V(1).Info("Canary deployment paused")
-							}
-							isPaused = true
-						}
-					}
-				}
+				continue
+			}
+
+			if !compareCurrentPodWithNewPod(params, pod, node) {
+				podsToDelete = append(podsToDelete, node)
+				continue
 			}
+
+			currentPods++
+			if podUtils.IsPodAvailable(pod, 0, metaNow) {
+				availablePods++
+			}
+			if podUtils.IsPodReady(pod) {
+				readyPods++
+			}
+
+			podsToCheckForRestarts = append(podsToCheckForRestarts, pod)
 		}
 	}
-	result.NewStatus = params.NewStatus.DeepCopy()
-	result.NewStatus.Status = string(ReplicaSetStatusCanary)
+
+	// Update result to reflect active pods currently experiencing restarts
+	// potentially placing canary into paused or failed state
+	manageCanaryPodRestarts(podsToCheckForRestarts, params, result)
+
+	// Update pod counts
 	result.NewStatus.Desired = desiredPods
 	result.NewStatus.Ready = readyPods
 	result.NewStatus.Available = availablePods
 	result.NewStatus.Current = currentPods
+
```

Comment:
I think we don't update the `result.NewStatus.Reason` with the info: `Paused`, `Failed`

Reply:
So while there is currently no `result.NewStatus.Reason` there is `result.NewStatus.Status` which is updated to:
- `canary` for active and paused canaries.
- `canary-failed` for failed canaries.

```
type ExtendedDaemonSetReplicaSetStatus struct {
	Status                   string `json:"status"`
	Desired                  int32  `json:"desired"`
	Current                  int32  `json:"current"`
	Ready                    int32  `json:"ready"`
	Available                int32  `json:"available"`
	IgnoredUnresponsiveNodes int32  `json:"ignoredUnresponsiveNodes"`
	// Conditions Represents the latest available observations of a DaemonSet's current state.
	// +listType=map
	// +listMapKey=type
	Conditions []ExtendedDaemonSetReplicaSetCondition `json:"conditions,omitempty"`
}
```

We interpret status as an enumerable field today (as opposed to a freeform description). Example where it is being taken into consideration as such: https://github.com/DataDog/extendeddaemonset/blob/master/controllers/extendeddaemonsetreplicaset/controller.go#L183